smartsim/_core/mli/infrastructure/storage/backbonefeaturestore.py

```diff
@@ -0,0 +1,399 @@
+            event_bytes = bytes(event)
+            self._event_buffer.append(event_bytes)
+        except Exception as ex:
+            raise ValueError("Unable to serialize event for sending") from ex
+
+    def _log_broadcast_start(self) -> None:
+        """Logs broadcast statistics"""
+        num_pending = len(self._event_buffer)
+        num_consumers = len(self._descriptors)
+        logger.debug(f"Broadcasting {num_pending} events to {num_consumers} consumers")
+
+    def _prune_unused_consumers(self) -> None:
+        """Performs maintenance on the channel cache by pruning any channel
+        that has been removed from the consumers list"""
+        active_consumers = set(self._descriptors)
+        current_channels = set(self._channel_cache.keys())
+
+        # find any cached channels that are now unused
+        inactive_channels = current_channels.difference(active_consumers)
+        new_channels = active_consumers.difference(current_channels)
+
+        for descriptor in inactive_channels:
+            self._channel_cache.pop(descriptor)
+
+        logger.debug(
+            f"Pruning {len(inactive_channels)} stale consumer channels"
+            f" and found {len(new_channels)} new channels"
+        )
+
+    def _get_comm_channel(self, descriptor: str) -> CommChannelBase:
+        """Helper method to build and cache a comm channel
+
+        :param descriptor: the descriptor to pass to the channel factory
+        :returns: the instantiated channel
+        :raises SmartSimError: if the channel fails to build"""
+        comm_channel = self._channel_cache[descriptor]
+        if comm_channel is not None:
+            return comm_channel
+
+        if self._channel_factory is None:
+            raise SmartSimError("No channel factory provided for consumers")
+
+        try:
+            channel = self._channel_factory(descriptor)
+            self._channel_cache[descriptor] = channel
+            return channel
+        except Exception as ex:
+            msg = f"Unable to construct channel with descriptor: {descriptor}"
+            logger.error(msg, exc_info=True)
+            raise SmartSimError(msg) from ex
+
+    def _broadcast(self) -> int:
+        """Broadcasts all buffered events to registered event consumers.
+
+        :return: the number of events broadcasted to consumers
+        :raises ValueError: if event serialization fails
+        :raises KeyError: if channel fails to attach using registered descriptors
+        :raises SmartSimError: if broadcasting fails"""
+
+        # allow descriptors to be empty since events are buffered
+        self._descriptors = set(x for x in self._backbone.notification_channels if x)
+        if not self._descriptors:
+            logger.warning("No event consumers are registered")
+            return 0
+
+        self._prune_unused_consumers()
+        self._log_broadcast_start()
+
+        num_sent: int = 0
+        next_event: t.Optional[bytes] = self._event_buffer.popleft()
```

Comment:
It looks like we're doing `popleft` for our next event, but also using `popleft` to discard events when the buffer is full. Is that what we want?

Reply:
The `deque` is a FIFO queue w/the left being the oldest item.  I believe we're good using this in both cases for the following reasons:

- we want to publish in order, `popleft` is good
- we want to discard the oldest item, `popleft` is good

Reply:
I see now, thanks for the clarification! I think there's something still nagging at me because the next event is either about to be broadcasted or discarded, which feels a bit weird.

Reply:
This might just be a me issue.

Reply:
Hang on I think I'm backwards on something.